Source/Core/Core/DSP/DSPAccelerator.h

```diff
@@ -14,41 +15,85 @@ class Accelerator
 public:
   virtual ~Accelerator() = default;
 
-  u16 Read(const s16* coefs);
+  u16 ReadSample(const s16* coefs);
   // Zelda ucode reads ARAM through 0xffd3.
-  u16 ReadD3();
-  void WriteD3(u16 value);
+  u16 ReadRaw();
+  void WriteRaw(u16 value);
 
   u32 GetStartAddress() const { return m_start_address; }
   u32 GetEndAddress() const { return m_end_address; }
   u32 GetCurrentAddress() const { return m_current_address; }
-  u16 GetSampleFormat() const { return m_sample_format; }
+  u16 GetSampleFormat() const { return m_sample_format.hex; }
+  s16 GetGain() const { return m_gain; }
   s16 GetYn1() const { return m_yn1; }
   s16 GetYn2() const { return m_yn2; }
   u16 GetPredScale() const { return m_pred_scale; }
+  u16 GetInput() const { return m_input; }
   void SetStartAddress(u32 address);
   void SetEndAddress(u32 address);
   void SetCurrentAddress(u32 address);
   void SetSampleFormat(u16 format);
+  void SetGain(s16 gain);
   void SetYn1(s16 yn1);
   void SetYn2(s16 yn2);
   void SetPredScale(u16 pred_scale);
+  void SetInput(u16 input);
 
   void DoState(PointerWrap& p);
 
 protected:
-  virtual void OnEndException() = 0;
+  virtual void OnRawReadEndException() = 0;
+  virtual void OnRawWriteEndException() = 0;
+  virtual void OnSampleReadEndException() = 0;
   virtual u8 ReadMemory(u32 address) = 0;
   virtual void WriteMemory(u32 address, u8 value) = 0;
+  u16 GetCurrentSample();
 
   // DSP accelerator registers.
   u32 m_start_address = 0;
   u32 m_end_address = 0;
   u32 m_current_address = 0;
-  u16 m_sample_format = 0;
+
+  enum class FormatSize : u16
+  {
+    Size4Bit = 0,
+    Size8Bit = 1,
+    Size16Bit = 2,
+    SizeInvalid = 3
+  };
+
+  enum class FormatDecode : u16
+  {
+    ADPCM = 0,        // ADPCM reads from ARAM, ACCA increments
+    MMIOPCMHalt = 1,  // PCM Reads from ACIN, ACCA doesn't increment
```

Comment:
I don't like "halt" here, which makes me think an exception is triggered or it does something like the `HALT` instruction. I think just calling it `MMIOPCM` would be clearer.